include/numerics/parsed_fem_function.h

```diff
@@ -48,238 +52,184 @@ namespace libMesh
 #else // LIBMESH_HAVE_FPARSER
   inline Output eval(char & libmesh_dbg_var(parser),
                      const std::string & libmesh_dbg_var(function_name),
-                     unsigned int libmesh_dbg_var(component_idx))
-  {
-    libmesh_error_msg("ERROR: This functionality requires fparser!");
-    return Output(0);
-  }
+                     unsigned int libmesh_dbg_var(component_idx)) const;
+#endif
+
+private:
+  const System& _sys;
+  std::string _expression;
+  std::vector<std::string> _subexpressions;
+  unsigned int _n_vars,
+    _n_requested_vars,
+    _n_requested_grad_components,
+    _n_requested_hess_components;
+  bool _requested_normals;
+#ifdef LIBMESH_HAVE_FPARSER
+  std::vector<FunctionParserBase<Output> > parsers;
+#else
+  std::vector<char> parsers;
 #endif
+  std::vector<Output> _spacetime;
 
+  // Flags for which variables need to be computed
 
-public:
+  // _need_var[v] is true iff value(v) is needed
+  std::vector<bool> _need_var;
 
-  /**
-   * Constructor.
-   */
-  explicit
-  ParsedFEMFunction (const System& sys,
-                     const std::string& expression,
-                     const std::vector<std::string>* additional_vars=NULL,
-                     const std::vector<Output>* initial_vals=NULL)
+  // _need_var_grad[v*LIBMESH_DIM+i] is true iff grad(v,i) is needed
+  std::vector<bool> _need_var_grad;
+
+#ifdef LIBMESH_ENABLE_SECOND_DERIVATIVES
+  // _need_var_hess[v*LIBMESH_DIM*LIBMESH_DIM+i*LIBMESH_DIM+j] is true
+  // iff grad(v,i,j) is needed
+  std::vector<bool> _need_var_hess;
+#endif // LIBMESH_ENABLE_SECOND_DERIVATIVES
+
+  // Variables/values that can be parsed and handled by the function parser
+  std::string variables;
+  std::vector<std::string> _additional_vars;
+  std::vector<Output> _initial_vals;
+};
+
+
+/*----------------------- Inline functions ----------------------------------*/
+
+template <typename Output>
+inline
+ParsedFEMFunction<Output>::ParsedFEMFunction
+  (const System& sys,
+   const std::string& expression,
+   const std::vector<std::string>* additional_vars,
+   const std::vector<Output>* initial_vals)
     : _sys(sys),
-      _expression(expression),
+      _expression (), // overridden by parse()
       _n_vars(sys.n_vars()),
       _n_requested_vars(0),
       _n_requested_grad_components(0),
       _n_requested_hess_components(0),
       _requested_normals(false),
       _need_var(_n_vars, false),
-      _need_var_grad(_n_vars*LIBMESH_DIM, false)
+      _need_var_grad(_n_vars*LIBMESH_DIM, false),
 #ifdef LIBMESH_ENABLE_SECOND_DERIVATIVES
-    , _need_var_hess(_n_vars*LIBMESH_DIM*LIBMESH_DIM, false)
+      _need_var_hess(_n_vars*LIBMESH_DIM*LIBMESH_DIM, false),
 #endif // LIBMESH_ENABLE_SECOND_DERIVATIVES
+      _additional_vars (additional_vars ? *additional_vars :
+                        std::vector<std::string>()),
+      _initial_vals (initial_vals ? *initial_vals :
+                     std::vector<Output>())
+  {
+    this->reparse(expression);
```

Comment:
Looks like `reparse()` is a non-virtual function, so this works fine, but I was working on some code the other day where I got bit by trying to call a virtual `init()` function from a constructor, and learned that the vtable stuff is not active in the constructor body, and with good reason...